sql/pgwire/server.go

```diff
@@ -167,21 +166,32 @@ func (s *Server) serveConn(conn net.Conn) error {
 		if err := buf.readUntypedMsg(conn); err != nil {
 			return err
 		}
-		version, rest = buf.msg[:4], buf.msg[4:]
+		version, err = buf.getInt32()
+		if err != nil {
+			return err
+		}
 	} else if !s.context.Insecure {
 		errSSLRequired = true
 	}
 
-	if bytes.Equal(version, version30) {
-		v3conn, err := newV3Conn(conn, rest, s.context.Executor)
-		if err != nil {
-			return err
-		}
+	if version == version30 {
+		v3conn := makeV3Conn(conn, s.context.Executor)
 		if errSSLRequired {
 			return v3conn.sendError(ErrSSLRequired)
 		}
-		return v3conn.serve()
+		if err := v3conn.parseOptions(buf.msg); err != nil {
+			return v3conn.sendError(err.Error())
+		}
+		if tlsConn, ok := conn.(*tls.Conn); ok {
+			tlsState := tlsConn.ConnectionState()
+			authenticationHook, err := security.UserAuthHook(s.context.Insecure, &tlsState)
+			if err != nil {
+				return v3conn.sendError(err.Error())
+			}
+			return v3conn.serve(authenticationHook)
```

Comment:
I thought about this a bit. Since authenticationHook needs to reach into `v3conn`'s internals (`opts.user`), its handling should be internal to `v3conn`. You might imagine a future in which we have a `conn` interface and several implementations - they might not share the authentication implementation.

Perhaps overthinking, but that's why.